lib/Dialect/FIRRTL/Transforms/LowerGroups.cpp

```diff
@@ -0,0 +1,321 @@
+    newModule.getBody().takeBody(group.getRegion());
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "      New Module: " << newModuleName << "\n";
+      llvm::dbgs() << "        ports:\n";
+      for (size_t i = 0, e = ports.size(); i != e; ++i) {
+        auto port = ports[i];
+        auto value = values[i];
+        llvm::dbgs() << "          - name: " << port.getName() << "\n"
+                     << "            type: " << port.type << "\n"
+                     << "            direction: " << port.direction << "\n"
+                     << "            value: " << value << "\n";
+      }
+    });
+
+    // Replace the original group with an instance.  Hook up the instance.
+    builder.setInsertionPointAfter(group);
+    auto moduleName = newModule.getModuleName();
+    auto instanceOp = builder.create<InstanceOp>(
+        group.getLoc(), /*moduleName=*/newModule,
+        /*name=*/
+        (Twine((char)tolower(moduleName[0])) + moduleName.drop_front()).str(),
+        NameKindEnum::DroppableName,
+        /*annotations=*/ArrayRef<Attribute>{},
+        /*portAnnotations=*/ArrayRef<Attribute>{}, /*lowerToBind=*/true);
+    instanceOp->setAttr("output_file",
+                        hw::OutputFileAttr::getFromFilename(
+                            builder.getContext(),
+                            "groups_" + circuitName + "_" + groupName + ".sv",
+                            /*excludeFromFileList=*/true));
+
+    for (unsigned portNum = 0, e = newModule.getNumPorts(); portNum < e;
+         ++portNum) {
+      if (instanceOp.getPortDirection(portNum) == Direction::In)
+        builder.create<StrictConnectOp>(builder.getUnknownLoc(),
+                                        instanceOp.getResult(portNum),
+                                        values[portNum]);
+      else {
+        OpBuilder::InsertionGuard guard(builder);
+        builder.setInsertionPointAfterValue(values[portNum]);
+        builder.create<StrictConnectOp>(
+            builder.getUnknownLoc(), values[portNum],
+            builder.create<RefResolveOp>(builder.getUnknownLoc(),
+                                         instanceOp.getResult(portNum)));
+      }
+    }
+    group.erase();
+
+    return WalkResult::advance();
+  });
+}
+
+/// Process a circuit to remove all groups in each module and top-level gropu
+/// declarations.
+void LowerGroupsPass::runOnOperation() {
+  LLVM_DEBUG(
+      llvm::dbgs() << "==----- Running LowerGroups "
+                      "-------------------------------------------------===\n");
+  CircuitOp circuitOp = getOperation();
+
+  // Initialize members which cannot be initialized automatically.
+  llvm::sys::SmartMutex<true> mutex;
+  circuitMutex = &mutex;
+  CircuitNamespace ns(circuitOp);
+  circuitNamespace = &ns;
+
+  // Lower the groups of each module.
+  SmallVector<FModuleOp> modules(circuitOp.getBodyBlock()->getOps<FModuleOp>());
+  llvm::parallelForEach(modules,
```

Comment:
Looks like there's a `mlir::parallelForEach` function that will check if the mlir context has multithreading enabled before dispatching multiple threads.